File: 프로그래머스/소수찾기.cpp
/*
	date : 2022.01.04
	title : 소수 찾기
	discribe : 에라토스테네스의 체 방법으로 O(nloglogn) 만에 모든 소수를 찾아보자
*/

#include <string>
#include <vector>
#include <math.h>
#define MAX 1000001

using namespace std;

// TIME OUT!!
// bool isPrime(int num) {
//     if (num % 2 == 0) return false;
    
//     for (int i = 3; i*i <= num; i++)
//         if (num % i == 0) 
//             return false;
// 	return true;
// }

int not_prime_nums[MAX] = {0, 1, 0, };

int solution(int n) {
    int answer = 0;
    
    for (int p = 2; p <= n; p++) {
        if (not_prime_nums[p] == 0) {
            for (int i = 2; p*i <= n; i++){
                not_prime_nums[p*i] = 1;
            }
        }
    }
    for (int i = 1; i <= n; i++) {
        if (not_prime_nums[i] == 0)
            answer++;
    }
    
    return answer;
}

File: 프로그래머스/Java/소수찾기.java
import java.util.ArrayList;
import java.util.HashSet;

class Solution {
    HashSet<Integer> hs = new HashSet<Integer>();
    //ArrayList<Integer> result = new ArrayList<Integer>(); 
    
    public int solution(String numbers) {
        
        String[] sNum = numbers.split("");      // "011" -> ["0", "1", "1"]
        int len = numbers.length();             // get length of input numbers
        boolean[] visited = new boolean[len];   // make visited boolean array
        String[] output = new String[len];      // make output string array
        
        // 뽑을 개수가 1부터 len까지 반복
        for (int i=1; i <= len; i++){
            //System.out.println("pick" + i);
            perm(sNum, output, visited, 0, len, i); // r = i, pick i
        }
        return hs.size();
        //return result.size();
    }
    
	// implement permutation algorithm using DFS
    public void perm(String[] arr, String[] output, boolean[] visited, int depth, int n, int r) {
        if (depth == r){
            print(output, r);
            return;
        }
        
        for (int i = 0; i < n; i++) {
            if (visited[i] != true) {
                // if not visited
                visited[i] = true;
                output[depth] = arr[i];
                perm(arr, output, visited, depth + 1, n, r);
                visited[i] = false;
            }
        }
    }
    
    
    public void print(String[] arr, int n) {
        String num = "";
        for (int i = 0; i < n; i++) {
            // System.out.print(arr[i] + " ");
            num += arr[i];
        }
        // System.out.println(num);
        int x = Integer.parseInt(num);
        // if (!result.contains(x) && isPrime(x)) {
        //     result.add(x);
        // }
		// using HashSet
		if (isPrime(x))
			hs.add(x); // 알아서 중복이 들어가지 않는다
    }
    
    static boolean isPrime(int x) {
        if (x < 2) return false;
        
        for (int i=2; i <= Math.sqrt(x); i++) {
            if (x % i == 0) {
                return false;
            }
        }
        return true;
    }
    
}